### DISCUSSION TITLE
Example of Vec<String> in Query?

### DISCUSSION THREAD

**Discussion by gd87429** [tokio-rs/axum#1719] [Q&A] [answered]
I looked at https://github.com/tokio-rs/axum/issues/434 but still have no idea how I'm supposed to make something like this work in 0.6.2:

```rust
#[derive(Deserialize)]
pub struct Q {
	a: Vec<String>,
}

pub async fn handler(
	Query(payload): Query<Q>,
) -> Result {
...
```

No matter if I try `?a=b&a=c` or `?a[]=b&a[]=c` or `?a=b,c` none of it works; I keep getting:

```
Failed to deserialize query string: invalid type: string "b", expected a sequence
```

Is there an example somewhere I'm missing?

**Comment by davidpdrsn** [comment, marked as answer]
There is no standard way to encode collections as query params, therefore `serde_url_encoded` doesn't support it. That is the parser axum's `Query` uses.

With that said you can use `axum_extra::extract::Query` instead which expects params the way HTML forms encodes them. Example here https://github.com/tokio-rs/axum/blob/main/axum-extra/src/extract/query.rs#L147. You can also use [`serde_qs`](https://docs.rs/serde_qs/0.11.0/serde_qs) which does support the `a[]` syntax.

**Reply by gd87429** [reply]
Ahh, that's what I was missing. I get it now. Thank you very much for explaining & taking the time!

**Comment by JuxhinDB** [comment]
While the answer works, I'm noticing some intermittent behaviour when an array of one item is passed. Take the following simplified example.

```rust

#[derive(Debug, Deserialize)]
pub(crate) Params {
     pub(crate): some_flag: Option<bool>,

     #[serde(flatten)]
     pub(crate): filters: Filters,
}

#[derive(Debug, Deserialize)]
pub(crate) Filters {
     #[serde(rename = "filter")]
     pub(crate) filters: Option<Vec<String>>,
}
```

Passing `?filter=foo&filter=bar` correctly parses this into `Some(["foo", "bar"])`. However just passing `?filter=foo` does not parse this into `Some(["foo"])`, instead it fails and returns `None`. Internal query rejection issues the same error:

```
Error { inner: Error("invalid type: string \"foo\", expected a sequence") })
```

Not sure if this is expected behaviour?

**Reply by davidpdrsn** [reply]
Does it work if you don't use `#[serde(flatten)]`?

**Reply by JuxhinDB** [reply]
Placing `Filters.filters` to the `Params` struct results in the same error and behaviour

**Reply by jplatte** [reply]
Serde's data model isn't exactly a great fit for query parameters. I think in this specific case it would help to change

```rust
#[serde(rename = "filter")]
pub(crate) filters: Option<Vec<String>>,
```

to

```rust
#[serde(default, rename = "filter")]
pub(crate) filters: Vec<String>,
```

**Reply by JuxhinDB** [reply]
~Thanks @jplatte! That seems to do the trick. Perhaps we can extend the answer to include this.~ 

Edit: apologies, turns out it _seemed_ like it was working due to the default. Still the same behaviour.

**Reply by JuxhinDB** [reply]
It might help to clarify that in my actual use-case the internal type if not `Vec<String>` but rather a custom type that impls `Deserialize`.

**Reply by jplatte** [reply]
Can you post a new issue on https://github.com/jplatte/serde_html_form please that fully reproduces the problem? If it's reproducible without `flatten` and with `default`, that would be useful.

**Reply by JuxhinDB** [reply]
I can confirm that this only happens when using `flatten`. I've opened the issue here https://github.com/jplatte/serde_html_form/issues/6 and have a PoC repo available here - https://github.com/JuxhinDB/serde-html-form-bug/blob/main/src/main.rs
